lib/plugins/aws/lib/monitorStack.js

```diff
@@ -5,19 +5,27 @@ const async = require('async');
 const chalk = require('chalk');
 
 module.exports = {
-  monitorStack(action, cfData, frequency) {
+  monitorStack(action, cfData, options = {}) {
     // Skip monitoring if stack was already created
     if (cfData === 'alreadyCreated') return BbPromise.bind(this).then(BbPromise.resolve());
 
-    // Monitor stack creation/update/removal
-    const validStatuses = ['CREATE_COMPLETE', 'UPDATE_COMPLETE', 'DELETE_COMPLETE'];
-    const loggedEvents = [];
     const region = this.provider.getRegion();
     const baseCfUrl = `https://${region}.console.aws.amazon.com/cloudformation/home`;
     const encodedStackId = `${encodeURIComponent(cfData.StackId)}`;
     const cfQueryString = `region=${region}#/stack/detail?stackId=${encodedStackId}`;
     const stackUrl = `${baseCfUrl}?${cfQueryString}`;
 
+    if (options.async) {
```

Comment:
So we still need have that issue with`Retain` in?

Is the deployments of those lambdas that slow? If we would do it via SDK, will we face the same problem?

Still, even it appears that we can't escape long deployments. I think not waiting for result is pretty controversial, as stack fails are not reported to the user.

Introducing that, we force the user to manually confirm with AWS console that stack deployment was successful (which is not very DX friendly). I think it'll be better to leave this long going process running, then user easily can confirm on a progress in a terminal. I would just output meaningful log as:

```
This deployment may take significant amount of time due to {reasons}.
If you want, you may stop monitoring of it with Ctrl+C (it won't stop the deployment on AWS side), and confirm later at {link} whether deployment finished successfully.
Note that no other deploy of this stack can be started until ongoing one finalizes.
```

Additional drawback of going _async_ is that going _async_ way, may leave impression that user can go on with updates and re-deploy shortly, which is not true. So I think long-going process makes a valuable message.

As a side note, an _async_ deployments make a nice use case for eventual native menu bar app, as it's the one that can monitor in background CF deployments and issue notifications when one finalizes (having that, `async` option will become more valid I think)

Reply:
Good points and I agree with most of them. I talked with @ganeshvlrk about this and we thought that it might be a good idea to do deployments and removals async since it takes on avg. ~20 minutes for a deployment to finish.

Ganesh had the idea that we can add a flag to await the deployment. Maybe we can invert this and provide a flag to do deployments async. This can be combined with your log message output proposal above.

> If we would do it via SDK, will we face the same problem?

I looked into support for Lambda@Edge via custom resources the other day and my conclusion was that it's not worth it since we would be dealing with the same drawbacks we're facing with CloudFormation deployments / removal.

> As a side note, an async deployments make a nice use case for eventual native menu bar app, as it's the one that can monitor in background CF deployments and issue notifications when one finalizes (having that, async option will become more valid I think)

I like that idea and i had a similar concept in my head when we worked on the EventGateway back then, but that's for another day... Although throwing an Electron app together which targets the use case you mentioned might be pretty simple to do.

Reply:
> Maybe we can invert this and provide a flag to do deployments async. This can be combined with your log message output proposal above.

I wonder whether in first place it'll be better to add support for handling Ctrl+C gently (and in certain (or all) cases suggesting user can stop monitoring that way)

As that way user will be definitely more aware of nature of the action.

In Node.js we can attach listener for exists done via Ctrl+C way.. so we probably can handle that nicely (like: abort naturally and display nice message with link where stack can be monitored). Not sure though if same can be done on Windows (that probably should be investigated before we go that route)

Then, still we may have use case for forcing that via config (e.g. user in all cases wants to abort early, so prefers it's just done for him).

I propose that we follow that route:

- Make this PR strictly about Lambda@Edge support (propagate eventual async support to another PR)
- (to be decided) In other PR if that's proven possible and reliable, introduce a way of gentle Ctrl+C handling
- (to be decided) In other PR, if we agree it's worthwhile, introduce an option that will generally allow _async_ deployments.

Reply:
Sounds like a good plan. I'll update this PR 👍